### PR TITLE
Add Gamification Rule for User Authentification

### PR DESCRIPTION
**Pull request by hbenali**

### CODE REVIEW

File: extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml [hunk 1]
```diff
@@ -142,6 +142,29 @@
...
+            <set-method>addPlugin</set-method>
+            <type>org.exoplatform.addons.gamification.service.setting.zone.model.ZoneConfig</type>
+            <init-params>
+                <value-param>
+                    <name>zone-name</name>
+                    <value>Attendence</value>
+                </value-param>
+                <value-param>
+                    <name>zone-description</name>
+                    <value>Attendence</value>
```
Comment (ahamdi) [inline extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml]:
```suggestion
                    <value>Attendance</value>
```

File: extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml [hunk 2]
```diff
@@ -142,6 +142,29 @@
...
+    <external-component-plugins>
+        <target-component>org.exoplatform.addons.gamification.service.setting.zone.ZoneRegistry</target-component>
+        <component-plugin>
+            <name>attendence.zone</name>
+            <set-method>addPlugin</set-method>
+            <type>org.exoplatform.addons.gamification.service.setting.zone.model.ZoneConfig</type>
+            <init-params>
+                <value-param>
+                    <name>zone-name</name>
+                    <value>Attendence</value>
```
Comment (ahamdi) [inline extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml]:
```suggestion
                    <value>Attendance</value>
```

File: extension/src/main/webapp/WEB-INF/conf/gamification/storage/rule/rules-storage-configuration.xml [hunk 3]
```diff
@@ -1278,4 +1278,38 @@
...
+                    <name>rule-event</name>
+                    <value>userLogin</value>
+                </value-param>
+                <value-param>
+                    <name>rule-score</name>
+                    <value>2</value>
+                </value-param>
+                <value-param>
+                    <name>rule-zone</name>
+                    <value>Attendence</value>
```
Comment (ahamdi) [inline extension/src/main/webapp/WEB-INF/conf/gamification/storage/rule/rules-storage-configuration.xml]:
```suggestion
                    <value>Attendance</value>
```

File: services/src/main/java/org/exoplatform/addons/gamification/GamificationConstant.java [hunk 4]
```diff
@@ -115,6 +115,12 @@
     /** UPLOAD AND SHARE DOCUMENT ON USER'S NETWORK CONTEXT */
     String GAMIFICATION_KNOWLEDGE_SHARE_UPLOAD__DOCUMENT_NETWORK_STREAM                           = "uploaddocumentOnNetworkStream";
 
+    /***************************************************************************************************************************************************/
+    /*************************************************** CONSTANTS TO MANAGE Attendence GAMIFICATION User Login **********************************************/
+    /***************************************************************************************************************************************************/
+    /** USER LOGIN CONTEXT */
+    String GAMIFICATION_ATTENDENCE_USER_LOGIN                                           = "userLogin";
```
Comment (ahamdi) [inline services/src/main/java/org/exoplatform/addons/gamification/GamificationConstant.java]:
```suggestion
    String GAMIFICATION_ATTENDANCE_USER_LOGIN                                           = "userLogin";
```

File: services/src/main/java/org/exoplatform/addons/gamification/listener/user/GamificationUserLoginListener.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        this.ruleService = ruleService;
+        this.identityManager = identityManager;
+        this.gamificationService = gamificationService;
+    }
+    @Override
+    public void onEvent(Event<ConversationRegistry, ConversationState> event) {
+        String username = event.getData().getIdentity().getUserId();
+        String sender = identityManager.getOrCreateIdentity(OrganizationIdentityProvider.NAME, username, false).getId();
+        gamificationService.createHistory(GAMIFICATION_ATTENDENCE_USER_LOGIN, sender, sender, null);
+        LOG.info("User Login Gamification for " + username);
```
Comment (ahamdi) [inline services/src/main/java/org/exoplatform/addons/gamification/listener/user/GamificationUserLoginListener.java]: Let's change log level to DEBUG
```suggestion
        LOG.debug("User Login Gamification for {}", username);
```

File: services/src/main/java/org/exoplatform/addons/gamification/listener/user/GamificationUserLoginListener.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+                                         GamificationService gamificationService) {
+        this.ruleService = ruleService;
+        this.identityManager = identityManager;
+        this.gamificationService = gamificationService;
+    }
+    @Override
+    public void onEvent(Event<ConversationRegistry, ConversationState> event) {
+        String username = event.getData().getIdentity().getUserId();
+        String sender = identityManager.getOrCreateIdentity(OrganizationIdentityProvider.NAME, username, false).getId();
+        gamificationService.createHistory(GAMIFICATION_ATTENDENCE_USER_LOGIN, sender, sender, null);
```
Comment (ahamdi) [inline services/src/main/java/org/exoplatform/addons/gamification/listener/user/GamificationUserLoginListener.java]:
```suggestion
        gamificationService.createHistory(GAMIFICATION_ATTENDANCE_USER_LOGIN, sender, sender, null);
```

File: services/src/main/resources/locale/addon/Gamification.properties [hunk 7]
```diff
@@ -12,6 +12,7 @@ exoplatform.gamification.leaderboard.domain.null=Overall
 exoplatform.gamification.leaderboard.domain.social=Social
 exoplatform.gamification.leaderboard.domain.knowledge=Knowledge
 exoplatform.gamification.leaderboard.domain.teamwork=Teamwork
+exoplatform.gamification.leaderboard.domain.attendence=Attendence
```
Comment (ahamdi) [inline services/src/main/resources/locale/addon/Gamification.properties]:
```suggestion
exoplatform.gamification.leaderboard.domain.attendance=Attendance
```

File: services/src/main/resources/locale/addon/Gamification.properties [hunk 8]
```diff
@@ -31,6 +32,7 @@ exoplatform.gamification.gamificationinformation.domain.Social=Social
 exoplatform.gamification.gamificationinformation.domain.Teamwork=Teamwork
 exoplatform.gamification.gamificationinformation.domain.Knowledge=Knowledge
 exoplatform.gamification.gamificationinformation.domain.Feedback=Feedback
+exoplatform.gamification.gamificationinformation.domain.Attendence=Attendence
```
Comment (ahamdi) [inline services/src/main/resources/locale/addon/Gamification.properties]:
```suggestion
exoplatform.gamification.gamificationinformation.domain.Attendance=Attendance
```

File: services/src/main/resources/locale/addon/Gamification.properties [hunk 9]
```diff
@@ -179,7 +184,9 @@ exoplatform.gamification.gamificationinformation.rule.description.def_completeTa
 exoplatform.gamification.gamificationinformation.rule.description.def_addForumTopic=You started a new topic in the forum
 exoplatform.gamification.gamificationinformation.rule.description.def_addActivityOnSpaceStream=You posted an activity
 exoplatform.gamification.gamificationinformation.rule.description.def_openNewTribeFeedback=You created a new tribe feedback
+exoplatform.gamification.gamificationinformation.rule.description.def_userLogin=You are logged in
 exoplatform.gamification.gamificationinformation.domain.Description.Social=Social
+exoplatform.gamification.gamificationinformation.domain.Description.Attendence=Attendence
```
Comment (ahamdi) [inline services/src/main/resources/locale/addon/Gamification.properties]:
```suggestion
exoplatform.gamification.gamificationinformation.domain.Description.Attendance=Attendance
```

File: services/src/main/resources/locale/addon/Gamification_en.properties [hunk 10]
```diff
@@ -11,6 +11,7 @@ exoplatform.gamification.leaderboard.title=Top users
 exoplatform.gamification.leaderboard.domain.null=Overall
 exoplatform.gamification.leaderboard.domain.social=Social
 exoplatform.gamification.leaderboard.domain.knowledge=Knowledge
+exoplatform.gamification.leaderboard.domain.attendence=Attendence
```
Comment (ahamdi) [inline services/src/main/resources/locale/addon/Gamification_en.properties]:
```suggestion
exoplatform.gamification.leaderboard.domain.attendance=Attendance
```

File: services/src/main/resources/locale/addon/Gamification_en.properties [hunk 11]
```diff
@@ -31,6 +32,7 @@ exoplatform.gamification.gamificationinformation.domain.Social=Social
 exoplatform.gamification.gamificationinformation.domain.Teamwork=Teamwork
 exoplatform.gamification.gamificationinformation.domain.Knowledge=Knowledge
 exoplatform.gamification.gamificationinformation.domain.Feedback=Feedback
+exoplatform.gamification.gamificationinformation.domain.Attendence=Attendence
```
Comment (ahamdi) [inline services/src/main/resources/locale/addon/Gamification_en.properties]:
```suggestion
exoplatform.gamification.gamificationinformation.domain.Attendance=Attendance
```

File: services/src/main/resources/locale/addon/Gamification_en.properties [hunk 12]
```diff
@@ -179,7 +184,9 @@ exoplatform.gamification.gamificationinformation.rule.description.def_completeTa
 exoplatform.gamification.gamificationinformation.rule.description.def_addForumTopic=You started a new topic in the forum
 exoplatform.gamification.gamificationinformation.rule.description.def_addActivityOnSpaceStream=You posted an activity
 exoplatform.gamification.gamificationinformation.rule.description.def_openNewTribeFeedback=You created a new tribe feedback
+exoplatform.gamification.gamificationinformation.rule.description.def_userLogin=You are logged in
 exoplatform.gamification.gamificationinformation.domain.Description.Social=Social
+exoplatform.gamification.gamificationinformation.domain.Description.Attendence=Attendence
```
Comment (ahamdi) [inline services/src/main/resources/locale/addon/Gamification_en.properties]:
```suggestion
exoplatform.gamification.gamificationinformation.domain.Description.Attendance=Attendance
```

File: services/src/main/resources/locale/addon/Gamification_fr.properties [hunk 13]
```diff
@@ -11,6 +11,7 @@ exoplatform.gamification.leaderboard.title=Top utilisateurs
 exoplatform.gamification.leaderboard.domain.null=Global
 exoplatform.gamification.leaderboard.domain.social=Social
 exoplatform.gamification.leaderboard.domain.knowledge=Science
+exoplatform.gamification.leaderboard.domain.attendence=Assiduit\u00E9
```
Comment (ahamdi) [inline services/src/main/resources/locale/addon/Gamification_fr.properties]:
```suggestion
exoplatform.gamification.leaderboard.domain.attendance=Assiduit\u00E9
```

File: services/src/main/resources/locale/addon/Gamification_fr.properties [hunk 14]
```diff
@@ -28,6 +29,7 @@ exoplatform.gamification.gamificationinformation.Event=\u00C9v\u00E9nement
 exoplatform.gamification.gamificationinformation.Date=Date
 exoplatform.gamification.gamificationinformation.Domain=Domaine
 exoplatform.gamification.gamificationinformation.domain.Social=Social
+exoplatform.gamification.gamificationinformation.domain.Attendence=Assiduit\u00E9
```
Comment (ahamdi) [inline services/src/main/resources/locale/addon/Gamification_fr.properties]:
```suggestion
exoplatform.gamification.gamificationinformation.domain.Attendance=Assiduit\u00E9
```

File: services/src/main/resources/locale/addon/Gamification_fr.properties [hunk 15]
```diff
@@ -179,7 +184,9 @@ exoplatform.gamification.gamificationinformation.rule.description.def_completeTa
 exoplatform.gamification.gamificationinformation.rule.description.def_addForumTopic=Vous avez commenc\u00E9 une nouvelle discussion dans le forum
 exoplatform.gamification.gamificationinformation.rule.description.def_addActivityOnSpaceStream=Vous avez post\u00E9 une activit\u00E9
 exoplatform.gamification.gamificationinformation.rule.description.def_openNewTribeFeedback=Vous avez ouvert un nouveau retour sur la Tribe
+exoplatform.gamification.gamificationinformation.rule.description.def_userLogin=Vous \u00EAtes connect\u00E9
 exoplatform.gamification.gamificationinformation.domain.Description.Social=Social
+exoplatform.gamification.gamificationinformation.domain.Description.Attendence=Assiduit\u00E9
```
Comment (ahamdi) [inline services/src/main/resources/locale/addon/Gamification_fr.properties]:
```suggestion
exoplatform.gamification.gamificationinformation.domain.Description.Attendance=Assiduit\u00E9
```

File: extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml [hunk 16]
```diff
@@ -142,6 +142,29 @@
         </component-plugin>
     </external-component-plugins>
 
+    <!-- Gamification zones : Attendence  zone -->
+    <external-component-plugins>
+        <target-component>org.exoplatform.addons.gamification.service.setting.zone.ZoneRegistry</target-component>
+        <component-plugin>
+            <name>attendence.zone</name>
```
Comment (ahamdi) [inline extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml]:
```suggestion
            <name>attendance.zone</name>
```

File: services/src/main/java/org/exoplatform/addons/gamification/GamificationConstant.java [hunk 17]
```diff
@@ -115,6 +115,12 @@
     /** UPLOAD AND SHARE DOCUMENT ON USER'S NETWORK CONTEXT */
     String GAMIFICATION_KNOWLEDGE_SHARE_UPLOAD__DOCUMENT_NETWORK_STREAM                           = "uploaddocumentOnNetworkStream";
 
+    /***************************************************************************************************************************************************/
+    /*************************************************** CONSTANTS TO MANAGE Attendence GAMIFICATION User Login **********************************************/
```
Comment (ahamdi) [inline services/src/main/java/org/exoplatform/addons/gamification/GamificationConstant.java]:
```suggestion
    /*************************************************** CONSTANTS TO MANAGE Attendance GAMIFICATION User Login **********************************************/
```

File: extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml [hunk 18]
```diff
@@ -142,6 +142,29 @@
         </component-plugin>
     </external-component-plugins>
 
+    <!-- Gamification zones : Attendence  zone -->
```
Comment (ahamdi) [inline extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml]:
```suggestion
    <!-- Gamification zones : Attendance  zone -->
```

File: services/src/main/java/org/exoplatform/addons/gamification/listener/user/GamificationUserLoginListener.java [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+import org.exoplatform.services.listener.Listener;
+import org.exoplatform.services.log.ExoLogger;
+import org.exoplatform.services.log.Log;
+import org.exoplatform.social.core.manager.IdentityManager;
+import org.exoplatform.services.security.ConversationRegistry;
+import org.exoplatform.services.security.ConversationState;
+import org.exoplatform.social.core.identity.provider.OrganizationIdentityProvider;
+
+
+import static org.exoplatform.addons.gamification.GamificationConstant.GAMIFICATION_ATTENDENCE_USER_LOGIN;
```
Comment (ahamdi) [inline services/src/main/java/org/exoplatform/addons/gamification/listener/user/GamificationUserLoginListener.java]:
```suggestion
import static org.exoplatform.addons.gamification.GamificationConstant.GAMIFICATION_ATTENDANCE_USER_LOGIN;
```

**Comment by rdenarie** [comment]
Does this listener launch if my session refresh but I have the remember me ? In this case, I do not "login" (enter login/password), but my session is refreshed ?
Should it ?

File: extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml (hunk 1 above)
Comment (hbenali) [reply extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml]: Done

File: extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml (hunk 2 above)
Comment (hbenali) [reply extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml]: Done

File: extension/src/main/webapp/WEB-INF/conf/gamification/storage/rule/rules-storage-configuration.xml (hunk 3 above)
Comment (hbenali) [reply extension/src/main/webapp/WEB-INF/conf/gamification/storage/rule/rules-storage-configuration.xml]: Done

File: services/src/main/java/org/exoplatform/addons/gamification/GamificationConstant.java (hunk 4 above)
Comment (hbenali) [reply services/src/main/java/org/exoplatform/addons/gamification/GamificationConstant.java]: Done

File: services/src/main/resources/locale/addon/Gamification_fr.properties (hunk 15 above)
Comment (hbenali) [reply services/src/main/resources/locale/addon/Gamification_fr.properties]: Done

File: extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml (hunk 16 above)
Comment (hbenali) [reply extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml]: Done

File: services/src/main/java/org/exoplatform/addons/gamification/listener/user/GamificationUserLoginListener.java (hunk 19 above)
Comment (hbenali) [reply services/src/main/java/org/exoplatform/addons/gamification/listener/user/GamificationUserLoginListener.java]: Done

File: services/src/main/java/org/exoplatform/addons/gamification/listener/user/GamificationUserLoginListener.java (hunk 6 above)
Comment (hbenali) [reply services/src/main/java/org/exoplatform/addons/gamification/listener/user/GamificationUserLoginListener.java]: Done

File: services/src/main/resources/locale/addon/Gamification.properties (hunk 8 above)
Comment (hbenali) [reply services/src/main/resources/locale/addon/Gamification.properties]: Done

File: extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml (hunk 18 above)
Comment (hbenali) [reply extension/src/main/webapp/WEB-INF/conf/gamification/storage/domain/domains-storage-configuration.xml]: Done

File: services/src/main/java/org/exoplatform/addons/gamification/GamificationConstant.java (hunk 17 above)
Comment (hbenali) [reply services/src/main/java/org/exoplatform/addons/gamification/GamificationConstant.java]: Done

File: services/src/main/java/org/exoplatform/addons/gamification/listener/user/GamificationUserLoginListener.java (hunk 5 above)
Comment (hbenali) [reply services/src/main/java/org/exoplatform/addons/gamification/listener/user/GamificationUserLoginListener.java]: Done

**Comment by plamarque** [comment]
> Does this listener launch if my session refresh but I have the remember me ? In this case, I do not "login" (enter login/password), but my session is refreshed ?
> Should it ?
Not sure what a session refresh is and when it is triggered. So I won't be able to answer directly. no matter how this thing works, here's the requirement for the gamification rule : Track how often I connect. It should count when I use remember me or an SSO.  It should count when I'm logged in twice (ore more times) at the same time on various browsers/devices.

**Comment by rdenarie** [comment]
> It should count when I use remember me or an SSO. 
It is what I want to know.

**Comment by hbenali** [comment]
It is already handled by the ConversationState event.

**Pull request merged**